### PR TITLE
Add image tags

### LINKED ISSUE #7: Image Tags
#### As a user I want to add tags to images.

**Story**:

Our users are loving the ability to add images to our service. However, now
that there are more than a handful of images, it is apparent that we need some
way to filter images by certain categories. In order to do that, we first need
to label images in some way. It is time to add the ability for our users to tag
images when they add them.

**Acceptance criteria**:
- [ ] When I add a new image, there is a form field to add tags.
- [ ] After I've saved an image with tags, I see those tags displayed with the
  associated image.
- [ ] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

**Reference material**:
- https://github.com/mbleigh/acts-as-taggable-on

**Dependencies**:
- Image Index

### PR DESCRIPTION
**Pull request by ozhe**
Closes #7

### CODE REVIEW

File: db/migrate/20161026213326_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# This migration comes from acts_as_taggable_on_engine (originally 1)
+class ActsAsTaggableOnMigration < ActiveRecord::Migration
+  def self.up
+    create_table :tags do |t|
+      t.string :name
```
Comment (shfung) [inline db/migrate/20161026213326_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Maybe we can add some not-null constraints to some of the columns?

File: app/views/images/new.html.erb [hunk 2]
```diff
@@ -1,5 +1,6 @@
 <p><%= link_to 'Home', root_path %></p>
 <%= simple_form_for @image do |f| %>
   <%= f.input :url %>
+  <%= f.input :tag_list, label: "Tags (separate with ',')" %>
```
Comment (shfung) [inline app/views/images/new.html.erb]: If you submit the form with an invalid URL and comma separated tags, are the tags still comma separated after you got back an error?

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -78,4 +78,14 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
     assert_response :ok
     assert_select 'img', count: 0
   end
+
+  test 'should add tags to image on creation' do
+    assert_difference('Image.count') do
+      post images_path, params: { image: { url: @image_url, tag_list: 'cat, dog' } }
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: Since you have added `tag_list` to your form, I think we should include the key `tag_list` to all the post request to the create action.

File: test/models/image_test.rb [hunk 4]
```diff
@@ -78,4 +78,12 @@ def test_index_order__starts_with_newest_image
 
     assert Image.index_order.each_cons(2).all? { |i, j| i.created_at >= j.created_at }, 'images in wrong order'
   end
+
+  def test_tag__can_be_added_to_image
+    image = Image.create!(url: 'https://www.google.com/test1.jpg')
```
Comment (shfung) [inline test/models/image_test.rb]: Maybe add a test to make sure an image can be created with tags?

File: db/migrate/20161026213326_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb (hunk 1 above)
Comment (shfung) [inline db/migrate/20161026213326_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Yes, even if the migrations are generated by a gem, it's okay to modify them to best accommodate our need.
And I guess you tested the provided interface and it won't save null or blank tag? While this is correct, there are many ways to get around that interface and save null or blank tag directly into the database.   It's generally a good practice to add not-null constraints to all columns except the ones that actually allow null.

File: db/migrate/20161026213326_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  def self.down
+    drop_table :taggings
+    drop_table :tags
+  end
+
+  def create_taggings
+    create_table :taggings do |t|
+      t.references :tag
```
Comment (shfung) [inline db/migrate/20161026213326_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: I think some more of these columns can have not-null constraints as well.

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -3,6 +3,9 @@
 class ImagesControllerTest < ActionDispatch::IntegrationTest
   setup do
     @image_url = 'https://www.appfolio.com/images/html/apm-logo-v2.png'
+    @tag_list = 'dog, cat'
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: Nice job adding `tag_list` key to all the `post` actions.  
Can we add one more test where `tag_list` is empty?
